sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -813,3 +815,66 @@ object RemoveLiteralFromGroupExpressions extends Rule[LogicalPlan] {
       a.copy(groupingExpressions = newGrouping)
   }
 }
+
+/**
+ * Remove the unnecessary evaluation from SortOrder, if they are already in child's output.
+ *
+ * As an example, "select a + 1, b from t order by a + 1" will be analyzed into:
+ * {{{
+ *   Project(reference#2, b#0,
+ *     Sort('a#1 + 1,
+ *       Project(('a + 1)#2, b#0, a#1),
+ *         Relation)
+ * }}}
+ * Then this rule can optimize it into:
+ * {{{
+ *   Project(reference#2, b#0,
+ *     Sort(reference#2,
+ *       Project(('a + 1)#2, b#0, a#1),
+ *         Relation)
+ * }}}
+ * Finally other optimize rules(column pruning, project collapse) can turn it into:
+ * {{{
+ *   Sort(reference#2,
+ *     Project(('a + 1)#2, b#0),
+ *       Relation)
+ * }}}
+ */
+object RemoveUnnecessarySortOrderEvaluation extends Rule[LogicalPlan] {
+  private def trimAliases(e: Expression) = e.transformUp {
```

Comment:
I would be good if we could just do this as part of CleanAliases in #7957.  Basically it seems that there is no reason to have an `Alias` except as a top level expression in Project or Aggregate.